rules/go_rules.build_defs

```diff
@@ -809,6 +809,143 @@ def cgo_test(name:str, srcs:list, data:list=None, deps:list=None, visibility:lis
         size = size,
     )
 
+def _go_install(name:str, get_roots:list, install:list, binary:bool, srcs:list, outs:list, deps:list, test_only:bool,
+               visibility:list, licences:list):
+    # Now compile it in a separate step.
+    install_cmd = ' '.join(install or [])
+    cmd = [
+        'export GOPATH="$(find \"$TMP_DIR\" \( -name src -o -name pkg \) -exec dirname {} \; | sort | uniq | tr \'\\n\' \':\' | sed \'s/:$//\')" GO111MODULE="off"',
+        '"$TOOLS_GO" install -gcflags "-trimpath \\\"$TMP_DIR\\\"" -asmflags "-trimpath \\\"$TMP_DIR\\\"" ' + install_cmd,
+        ]
+    if CONFIG.GOROOT:
+        cmd = [f'export GOROOT={CONFIG.GOROOT}'] + cmd
+    if package_name():
+        cmd += [
+            # The outputs tend to end up in subdirectories (Go seems to match them to the location the source came from)
+            'rm -rf bin' if binary else 'rm -rf pkg',
+            'mv "$PKG_DIR/bin" .' if binary else 'mv "$PKG_DIR/pkg" .',
+        ]
+
+    if binary:
+        outs = ['bin/' + name]
+    else:
+        outs = [f'pkg/{CONFIG.GOOS}_{CONFIG.GOARCH}/{out}' for out in outs]
+        # Outputs are created one directory down from where we want them.
+        # For most it doesn't matter but the top-level one will get lost.
+        pkg_get_roots = [f'pkg/{CONFIG.GOOS}_{CONFIG.GOARCH}/{getroot}' for getroot in get_roots]
+        cmd += [f' if [ -f {out}.a ]; then mkdir -p {out} && cp {out}.a {out}; fi' for out in pkg_get_roots]
+    return build_rule(
+        name = name,
+        tag = "a_rule" if not binary else None,
+        outs = outs,
+        deps = deps,
+        srcs = srcs,
+        tools = {
+            'go': [CONFIG.GO_TOOL],
+        },
+        visibility = visibility,
+        building_description = 'Compiling...',
+        cmd = ' && '.join(cmd),
+        binary = binary,
+        requires = ['go', 'go_src'],
+        test_only = test_only,
+        labels = ['link:plz-out/go'],
+        sandbox = False,
+        needs_transitive_deps = True,
+        licences = licences,
+    )
+
+def go_mod_download(name:str, module:str, version:str, test_only:bool=False, visibility:list=None,
+                    licences:list=None, hashes:list=None, labels:list=[],_tag:str=''):
+    cmds = [
+        "export GO111MODULE=on && export TMPDIR=$TMPDIR/tmp && mkdir -p $TMPDIR",
+        "$TOOL mod init dummy", # Create a dummy module
+        # TODO(jpoole): I don't think this is portable beyond modern GNU versions of grep. Might need to switch to sed/awk for this
+        f"export MOD_DIR=$($TOOL mod download -x -modcacherw -json {module}@{version} | grep -Po '\"Dir\": \"\K([^\"]+)')",
```

Comment:
Could use `awk -F\" '/"Dir": / { print $4 }'` or `grep -F '"Dir": ' | cut -d\" -f4` here instead of the PCRE regex in `grep` - both are POSIX-compliant.

Reply:
Ah brilliant thanks for looking into that for me :)